# Project data processing

In [2]:
import numpy as np
import pandas as pd
import glob
import os

In [3]:
path='d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\data\ '  # make your own path to the data
csv_path=path+"*.csv"
files=glob.glob(csv_path)      # gives you list of csv files in the directory, unfortunately sorted in a wrong way
numbers=[i for i in range(len(files))]
corrected_files=[files[i][:63]+str(numbers[i])+".csv" for i in range(len(files)) ]

In [4]:
list_of_DFs = [pd.read_csv(file) for file in corrected_files]

In [5]:
len(list_of_DFs)

10920

In [6]:
path2='d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\data2\ '  # make your own path to the data
csv_path2=path2+"*.csv"
files2=glob.glob(csv_path2)      # gives you list of csv files in the directory, unfortunately sorted in a wrong way
numbers2=[i for i in range(len(files2))]
corrected_files2=[files2[i][:64]+str(numbers2[i])+".csv" for i in range(len(files2)) ]

In [7]:
list_of_DFs2 = [pd.read_csv(file) for file in corrected_files2]

In [8]:
len(list_of_DFs2)

10920

In [9]:
list_of_DFs[10]

city  year     month  Giorno  T Media    T min    T max Precip. Umidità  \
0   Bari  1985  Novembre       1  17.3 °C    14 °C  22.2 °C       -    78 %   
1   Bari  1985  Novembre       2  16.4 °C    12 °C  21.2 °C    2 mm    74 %   
2   Bari  1985  Novembre       3  15.7 °C    11 °C  20.4 °C     n/d    70 %   
3   Bari  1985  Novembre       4  15.9 °C    14 °C  19.4 °C   21 mm    86 %   
4   Bari  1985  Novembre       5  16.2 °C    11 °C  19.8 °C    2 mm    76 %   
5   Bari  1985  Novembre       6  18.8 °C    15 °C  21.4 °C    5 mm    82 %   
6   Bari  1985  Novembre       7  15.2 °C    10 °C    17 °C    3 mm    63 %   
7   Bari  1985  Novembre       8  13.6 °C   8.8 °C  18.4 °C       -    65 %   
8   Bari  1985  Novembre       9  14.3 °C    10 °C  18.6 °C       -    81 %   
9   Bari  1985  Novembre      10  18.4 °C    12 °C  26.2 °C       -    73 %   
10  Bari  1985  Novembre      11  19.3 °C  15.2 °C  26.4 °C       -    77 %   
11  Bari  1985  Novembre      12  19.7 °C  15.8 °C    24 °C       -    75 %   
12  Bari  1985  Novembre      13  20.6 °C    18 °C  25.2 °C       -    76 %   
13  Bari  1985  Novembre      14  16.6 °C    11 °C    21 °C       -    64 %   
14  Bari  1985  Novembre      15  12.5 °C     8 °C  15.6 °C       -    67 %   
15  Bari  1985  Novembre      16  10.9 °C   7.6 °C  13.8 °C     n/d    75 %   
16  Bari  1985  Novembre      17  10.6 °C    10 °C    12 °C    7 mm    87 %   
17  Bari  1985  Novembre      18  12.1 °C    10 °C  14.2 °C    4 mm    89 %   
18  Bari  1985  Novembre      19  11.8 °C     9 °C  13.4 °C   33 mm    97 %   
19  Bari  1985  Novembre      20  13.3 °C    10 °C    18 °C    3 mm    85 %   
20  Bari  1985  Novembre      21  12.1 °C     7 °C    15 °C       -    76 %   
21  Bari  1985  Novembre      22  11.5 °C     7 °C  16.2 °C     n/d    80 %   
22  Bari  1985  Novembre      23  10.6 °C     7 °C  14.2 °C     n/d    94 %   
23  Bari  1985  Novembre      24   9.1 °C     4 °C    13 °C       -    91 %   
24  Bari  1985  Novembre      25  11.1 °C     6 °C    15 °C     n/d    83 %   
25  Bari  1985  Novembre      26  11.1 °C     9 °C    15 °C    5 mm    90 %   
26  Bari  1985  Novembre      27  10.3 °C   7.6 °C  13.4 °C    4 mm    81 %   
27  Bari  1985  Novembre      28  10.2 °C   9.6 °C  11.2 °C    1 mm    92 %   
28  Bari  1985  Novembre      29   9.7 °C     5 °C  13.4 °C    1 mm    82 %   
29  Bari  1985  Novembre      30  11.1 °C    -5 °C  18.2 °C       -    79 %   

    Vento Max  Raffica Fenomeni  Info  
0   16.5 km/h        -  Nessuno   NaN  
1   33.5 km/h        -      NaN   NaN  
2   14.8 km/h        -      NaN   NaN  
3   22.2 km/h  47 km/h      NaN   NaN  
4   16.5 km/h        -  Nessuno   NaN  
5     37 km/h        -      NaN   NaN  
6   24.1 km/h        -      NaN   NaN  
7   11.1 km/h        -  Nessuno   NaN  
8   11.1 km/h        -  Nessuno   NaN  
9   14.8 km/h        -  Nessuno   NaN  
10  29.4 km/h        -  Nessuno   NaN  
11  22.2 km/h        -  Nessuno   NaN  
12  27.8 km/h        -  Nessuno   NaN  
13  27.8 km/h        -  Nessuno   NaN  
14  18.3 km/h        -  Nessuno   NaN  
15    13 km/h        -      NaN   NaN  
16  18.3 km/h        -      NaN   NaN  
17  25.9 km/h        -      NaN   NaN  
18  14.8 km/h        -      NaN   NaN  
19  25.9 km/h        -      NaN   NaN  
20  24.1 km/h        -  Nessuno   NaN  
21  18.3 km/h        -      NaN   NaN  
22  22.2 km/h        -      NaN   NaN  
23  27.8 km/h  48 km/h  Nessuno   NaN  
24  25.9 km/h        -      NaN   NaN  
25  29.4 km/h        -      NaN   NaN  
26  24.1 km/h        -      NaN   NaN  
27  20.6 km/h        -      NaN   NaN  
28  27.8 km/h        -      NaN   NaN  
29  16.5 km/h        -      NaN   NaN

In [10]:
list_of_DFs2[10]

city  year     month  Giorno  T Media    T min    T max Precip.  \
0   Bergamo  1985  Novembre       1    12 °C    11 °C  14.4 °C    1 mm   
1   Bergamo  1985  Novembre       2  11.3 °C     6 °C    15 °C       -   
2   Bergamo  1985  Novembre       3   9.7 °C     5 °C  14.4 °C       -   
3   Bergamo  1985  Novembre       4   9.8 °C   4.6 °C  14.2 °C       -   
4   Bergamo  1985  Novembre       5   9.3 °C   7.8 °C  10.4 °C    1 mm   
5   Bergamo  1985  Novembre       6  12.2 °C   6.8 °C  18.4 °C    3 mm   
6   Bergamo  1985  Novembre       7   9.9 °C   3.8 °C  14.8 °C       -   
7   Bergamo  1985  Novembre       8   8.4 °C   3.2 °C  12.4 °C       -   
8   Bergamo  1985  Novembre       9   8.1 °C   4.8 °C  10.8 °C     n/d   
9   Bergamo  1985  Novembre      10   8.8 °C   7.8 °C   9.4 °C    2 mm   
10  Bergamo  1985  Novembre      11     9 °C     8 °C  11.8 °C       -   
11  Bergamo  1985  Novembre      12   6.9 °C   5.3 °C   7.6 °C   20 mm   
12  Bergamo  1985  Novembre      13   7.4 °C     6 °C  10.4 °C   16 mm   
13  Bergamo  1985  Novembre      14   5.1 °C   0.8 °C    11 °C       -   
14  Bergamo  1985  Novembre      15   3.3 °C    -1 °C   9.2 °C       -   
15  Bergamo  1985  Novembre      16   1.2 °C    -1 °C   5.4 °C       -   
16  Bergamo  1985  Novembre      17   1.8 °C    -3 °C     6 °C       -   
17  Bergamo  1985  Novembre      18   1.3 °C    -1 °C   3.6 °C       -   
18  Bergamo  1985  Novembre      19   2.3 °C   0.6 °C     4 °C       -   
19  Bergamo  1985  Novembre      20   2.7 °C   0.2 °C   4.2 °C    8 mm   
20  Bergamo  1985  Novembre      21   2.6 °C   0.8 °C   3.3 °C   11 mm   
21  Bergamo  1985  Novembre      22   3.1 °C   1.6 °C     4 °C    2 mm   
22  Bergamo  1985  Novembre      23   4.3 °C   2.6 °C   5.2 °C   16 mm   
23  Bergamo  1985  Novembre      24   5.3 °C     3 °C   9.3 °C    9 mm   
24  Bergamo  1985  Novembre      25   3.7 °C  -1.2 °C   8.2 °C       -   
25  Bergamo  1985  Novembre      26   3.8 °C  -0.2 °C   7.8 °C       -   
26  Bergamo  1985  Novembre      27   2.7 °C  -0.4 °C   5.8 °C       -   
27  Bergamo  1985  Novembre      28   1.3 °C    -2 °C     6 °C       -   
28  Bergamo  1985  Novembre      29   0.9 °C    -3 °C   6.6 °C       -   
29  Bergamo  1985  Novembre      30   3.1 °C    -1 °C   5.1 °C       -   

   Umidità  Vento Max Raffica Fenomeni  Info  
0     88 %   9.4 km/h       -      NaN   NaN  
1     69 %  11.1 km/h       -      NaN   NaN  
2     72 %   5.4 km/h       -  Nessuno   NaN  
3     67 %   7.6 km/h       -  Nessuno   NaN  
4     85 %          -       -      NaN   NaN  
5     62 %  33.5 km/h       -  Nessuno   NaN  
6     71 %    13 km/h       -  Nessuno   NaN  
7     75 %   5.4 km/h       -  Nessuno   NaN  
8     85 %          -       -      NaN   NaN  
9     90 %          -       -      NaN   NaN  
10    89 %          -       -      NaN   NaN  
11    86 %   5.4 km/h       -      NaN   NaN  
12    78 %  22.2 km/h       -      NaN   NaN  
13    57 %  11.1 km/h       -  Nessuno   NaN  
14    65 %   7.6 km/h       -  Nessuno   NaN  
15    85 %          -       -  Nessuno   NaN  
16    78 %   7.6 km/h       -  Nessuno   NaN  
17    80 %  11.1 km/h       -  Nessuno   NaN  
18    76 %   5.4 km/h       -      NaN   NaN  
19    84 %   7.6 km/h       -      NaN   NaN  
20    88 %          -       -      NaN   NaN  
21    85 %          -       -      NaN   NaN  
22    90 %   9.4 km/h       -      NaN   NaN  
23    86 %   9.4 km/h       -  Nessuno   NaN  
24    84 %   7.6 km/h       -      NaN   NaN  
25    77 %  11.1 km/h       -  Nessuno   NaN  
26    83 %          -       -  Nessuno   NaN  
27    86 %   5.4 km/h       -      NaN   NaN  
28    77 %   9.4 km/h       -      NaN   NaN  
29    79 %          -       -  Nessuno   NaN

In [11]:
# deleting columns which are not needed
col_to_del=["Raffica","Fenomeni","Info"]
for table in list_of_DFs:
    for x in col_to_del:
        del table[x]

for table in list_of_DFs2:
    for x in col_to_del:
        del table[x]

In [12]:
'''  For making computation faster, first let joins appropriate months into quarters  '''
quarterly_data=[]
for i in range(int(len(list_of_DFs)/3)):
    quarterly_data.append(pd.concat(list_of_DFs[i*3:i*3+3]))
    if i%500==0:
        print(i)
        
for i in range(int(len(list_of_DFs2)/3)):
    quarterly_data.append(pd.concat(list_of_DFs2[i*3:i*3+3]))
    if i%500==0:
        print(i)

0
500
1000
1500
2000
2500
3000
3500
0
500
1000
1500
2000
2500
3000
3500


In [13]:
''' Making lists of quarterly dataframes, for convenience using copy of the original dataframe 
in order to have more recent reference ''' 
first_q=[x.copy(deep=True) for x in quarterly_data[::4]]
second_q=[x.copy(deep=True) for x in quarterly_data[1::4]]
third_q=[x.copy(deep=True) for x in quarterly_data[2::4]]
fourth_q=[x.copy(deep=True) for x in quarterly_data[3::4]]

In [15]:
import re

In [16]:
'''  Converting strings to numbers where needed, for time efficiency step by step  '''
for t in range(len(first_q)):
    if t%100==0:
        print(t)
    table=first_q[t]
    for i in range(4,10):
        for j in range(len(table.iloc[:,i])):
            y=table.iloc[j,i].split(".")
            z=".".join([re.sub("[^0-9]"," ",x)  for x in y])
            try:
                table.iloc[j,i]=float(z)
            except:
                table.iloc[j,i]=z
            

0
20
40
60
80
100
120
140
160
180
200
220
240
260
280
300
320
340
360
380
400
420
440
460
480
500
520
540
560
580
600
620
640
660
680
700
720
740
760
780
800
820
840
860
880
900
920
940
960
980
1000
1020
1040
1060
1080
1100
1120
1140
1160
1180
1200
1220
1240
1260
1280
1300
1320
1340
1360
1380
1400
1420
1440
1460
1480
1500
1520
1540
1560
1580
1600
1620
1640
1660
1680
1700
1720
1740
1760
1780
1800


In [74]:
'''     Exporting quarterly dataframes to the csv files        '''
qs_paths=['d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\First_q\ '+str(10000+A)+".csv" for A in range(len(first_q))]
[first_q[i].to_csv(qs_paths[i], index = None, header=True) for i in range(len(first_q))]

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

In [ ]:
for t in range(len(second_q)):
    if t%100==0:
        print(t)
    table=second_q[t]
    for i in range(4,10):
        for j in range(len(table.iloc[:,i])):
            y=table.iloc[j,i].split(".")
            z=".".join([re.sub("[^0-9]"," ",x)  for x in y])
            try:
                table.iloc[j,i]=float(z)
            except:
                table.iloc[j,i]=z

In [75]:
'''     Exporting quarterly dataframes to the csv files        '''
qs_paths2=['d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\second_q\ '+str(10000+A)+".csv" for A in range(len(second_q))]
[second_q[i].to_csv(qs_paths2[i], index = None, header=True) for i in range(len(second_q))]

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

In [ ]:
for t in range(len(third_q)):
    if t%100==0:
        print(t)
    table=third_q[t]
    for i in range(4,10):
        for j in range(len(table.iloc[:,i])):
            y=table.iloc[j,i].split(".")
            z=".".join([re.sub("[^0-9]"," ",x)  for x in y])
            try:
                table.iloc[j,i]=float(z)
            except:
                table.iloc[j,i]=z

In [78]:
'''     Exporting quarterly dataframes to the csv files        '''
qs_paths3=['d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\Third_q\ '+str(10000+A)+".csv" for A in range(len(third_q))]
[third_q[i].to_csv(qs_paths3[i], index = None, header=True) for i in range(len(third_q))]

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

In [ ]:
for t in range(len(fourth_q)):
    if t%100==0:
        print(t)
    table=fourth_q[t]
    for i in range(4,10):
        for j in range(len(table.iloc[:,i])):
            y=table.iloc[j,i].split(".")
            z=".".join([re.sub("[^0-9]"," ",x)  for x in y])
            try:
                table.iloc[j,i]=float(z)
            except:
                table.iloc[j,i]=z

In [81]:
'''     Exporting quarterly dataframes to the csv files        '''
qs_paths4=['d:\moje_dokumenty\Desktop\IES\semester 11\Python\project\Fourth_q\ '+str(10000+A)+".csv" for A in range(len(fourth_q))]
[fourth_q[i].to_csv(qs_paths4[i], index = None, header=True) for i in range(len(fourth_q))]

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

In [14]:
first_q[0]

city  year    month  Giorno  T Media    T min    T max Precip. Umidità  \
0   Bari  1985  Gennaio       1   6.6 °C   5.6 °C     8 °C    1 mm    89 %   
1   Bari  1985  Gennaio       2   5.7 °C     3 °C   9.6 °C    1 mm    91 %   
2   Bari  1985  Gennaio       3   4.2 °C     2 °C     8 °C       -    85 %   
3   Bari  1985  Gennaio       4   4.2 °C   0.8 °C     8 °C       -    77 %   
4   Bari  1985  Gennaio       5   7.7 °C   3.4 °C  15.2 °C       -    81 %   
5   Bari  1985  Gennaio       6   1.7 °C    -1 °C   4.4 °C   11 mm    94 %   
6   Bari  1985  Gennaio       7   2.4 °C  -1.2 °C     6 °C    1 mm    85 %   
7   Bari  1985  Gennaio       8   1.5 °C   0.6 °C     3 °C    3 mm    92 %   
8   Bari  1985  Gennaio       9   3.4 °C  -0.6 °C   7.4 °C       -    85 %   
9   Bari  1985  Gennaio      10   0.2 °C    -1 °C   2.2 °C       -    68 %   
10  Bari  1985  Gennaio      11   1.1 °C    -2 °C   4.4 °C       -    80 %   
11  Bari  1985  Gennaio      12   1.3 °C    -3 °C   5.6 °C       -    71 %   
12  Bari  1985  Gennaio      13   3.9 °C  -3.2 °C   9.2 °C       -    73 %   
13  Bari  1985  Gennaio      14   6.4 °C     2 °C    13 °C       -    88 %   
14  Bari  1985  Gennaio      15     9 °C     5 °C  12.2 °C     n/d    90 %   
15  Bari  1985  Gennaio      16   9.8 °C   8.4 °C  11.4 °C   12 mm    86 %   
16  Bari  1985  Gennaio      17   9.4 °C     6 °C  11.2 °C       -    76 %   
17  Bari  1985  Gennaio      18     8 °C   5.8 °C    11 °C     n/d    85 %   
18  Bari  1985  Gennaio      19   9.4 °C     7 °C  14.2 °C    4 mm    89 %   
19  Bari  1985  Gennaio      20   8.1 °C     6 °C    12 °C       -    89 %   
20  Bari  1985  Gennaio      21   9.8 °C   4.2 °C  16.2 °C       -    86 %   
21  Bari  1985  Gennaio      22  13.2 °C    10 °C  17.6 °C       -    89 %   
22  Bari  1985  Gennaio      23  13.8 °C    12 °C    17 °C       -    88 %   
23  Bari  1985  Gennaio      24  14.9 °C    10 °C    21 °C     n/d    62 %   
24  Bari  1985  Gennaio      25  10.1 °C     6 °C  13.2 °C    1 mm    79 %   
25  Bari  1985  Gennaio      26   9.4 °C     4 °C  15.8 °C       -    73 %   
26  Bari  1985  Gennaio      27   8.9 °C     6 °C    11 °C   11 mm    90 %   
27  Bari  1985  Gennaio      28   9.5 °C     8 °C    11 °C    9 mm    94 %   
28  Bari  1985  Gennaio      29   9.3 °C     9 °C  10.4 °C   12 mm    90 %   
29  Bari  1985  Gennaio      30   8.2 °C     5 °C  12.2 °C       -    87 %   
..   ...   ...      ...     ...      ...      ...      ...     ...     ...   
1   Bari  1985    Marzo       2  12.2 °C     6 °C    20 °C     n/d    82 %   
2   Bari  1985    Marzo       3  12.3 °C     7 °C    17 °C    1 mm    77 %   
3   Bari  1985    Marzo       4   9.8 °C     6 °C    14 °C       -    88 %   
4   Bari  1985    Marzo       5  12.1 °C   6.4 °C    18 °C       -    79 %   
5   Bari  1985    Marzo       6  13.1 °C    10 °C  18.2 °C       -    79 %   
6   Bari  1985    Marzo       7  14.1 °C  10.4 °C    18 °C       -    79 %   
7   Bari  1985    Marzo       8  13.7 °C    11 °C  16.6 °C       -    83 %   
8   Bari  1985    Marzo       9  12.3 °C    11 °C  15.2 °C    4 mm    91 %   
9   Bari  1985    Marzo      10  10.6 °C     9 °C    12 °C     n/d    92 %   
10  Bari  1985    Marzo      11   8.6 °C     7 °C    10 °C    5 mm    93 %   
11  Bari  1985    Marzo      12   8.2 °C   6.8 °C    10 °C    5 mm    93 %   
12  Bari  1985    Marzo      13  10.8 °C     9 °C    13 °C    1 mm    91 %   
13  Bari  1985    Marzo      14  10.9 °C     9 °C  14.8 °C    5 mm    86 %   
14  Bari  1985    Marzo      15  11.1 °C     8 °C  14.8 °C       -    83 %   
15  Bari  1985    Marzo      16  10.9 °C     8 °C  16.2 °C    1 mm    80 %   
16  Bari  1985    Marzo      17   9.6 °C     5 °C    15 °C    1 mm    79 %   
17  Bari  1985    Marzo      18   8.3 °C   4.4 °C  14.6 °C    1 mm    79 %   
18  Bari  1985    Marzo      19   6.6 °C     1 °C  11.2 °C       -    77 %   
19  Bari  1985    Marzo      20   9.4 °C     3 °C    15 °C     n/d    68 %   
20  Bari  1985    Marzo   

In [406]:
first_q[3]

city  year    month  Giorno  T Media    T min    T max Precip. Umidità  \
0   Bari  1988  Gennaio       1   8.9 °C     8 °C  10.4 °C       -    92 %   
1   Bari  1988  Gennaio       2   9.6 °C     5 °C    14 °C       -    86 %   
2   Bari  1988  Gennaio       3  10.6 °C   4.6 °C    16 °C       -    75 %   
3   Bari  1988  Gennaio       4  11.3 °C    10 °C    13 °C   11 mm    91 %   
4   Bari  1988  Gennaio       5   9.9 °C     6 °C  14.6 °C       -    87 %   
5   Bari  1988  Gennaio       6  11.3 °C     6 °C    19 °C       -    73 %   
6   Bari  1988  Gennaio       7  11.9 °C     7 °C  18.2 °C       -    71 %   
7   Bari  1988  Gennaio       8   9.6 °C     8 °C    12 °C    1 mm    92 %   
8   Bari  1988  Gennaio       9   9.2 °C     8 °C  11.2 °C   10 mm    81 %   
9   Bari  1988  Gennaio      10   8.1 °C   4.8 °C    11 °C    8 mm    77 %   
10  Bari  1988  Gennaio      11   5.8 °C     3 °C  10.4 °C       -    80 %   
11  Bari  1988  Gennaio      12   7.2 °C   3.6 °C    12 °C       -    82 %   
12  Bari  1988  Gennaio      13   7.1 °C     3 °C  10.4 °C       -    86 %   
13  Bari  1988  Gennaio      14   8.8 °C     6 °C    13 °C       -    85 %   
14  Bari  1988  Gennaio      15   9.8 °C     6 °C    14 °C     n/d    87 %   
15  Bari  1988  Gennaio      16   9.6 °C     6 °C    13 °C       -    89 %   
16  Bari  1988  Gennaio      17   9.4 °C     6 °C  15.3 °C       -    86 %   
17  Bari  1988  Gennaio      18   8.7 °C   4.4 °C    12 °C       -    91 %   
18  Bari  1988  Gennaio      19  10.9 °C   7.4 °C  13.8 °C    7 mm    90 %   
19  Bari  1988  Gennaio      20   9.8 °C     7 °C  12.4 °C    3 mm    92 %   
20  Bari  1988  Gennaio      21   9.6 °C     6 °C  14.8 °C     n/d    89 %   
21  Bari  1988  Gennaio      22   9.3 °C     8 °C    18 °C    3 mm    87 %   
22  Bari  1988  Gennaio      23   8.6 °C     6 °C  12.6 °C   12 mm    88 %   
23  Bari  1988  Gennaio      24   6.9 °C     4 °C  10.6 °C    6 mm    78 %   
24  Bari  1988  Gennaio      25   8.4 °C     1 °C    14 °C       -    74 %   
25  Bari  1988  Gennaio      26  13.1 °C     8 °C    18 °C    1 mm    68 %   
26  Bari  1988  Gennaio      27  11.6 °C     7 °C    17 °C       -    79 %   
27  Bari  1988  Gennaio      28  11.4 °C     7 °C  15.8 °C       -    89 %   
28  Bari  1988  Gennaio      29  13.3 °C  10.6 °C    19 °C       -    74 %   
29  Bari  1988  Gennaio      30  12.9 °C     8 °C    17 °C     n/d    59 %   
..   ...   ...      ...     ...      ...      ...      ...     ...     ...   
1   Bari  1988    Marzo       2   5.3 °C   1.8 °C     9 °C       -    65 %   
2   Bari  1988    Marzo       3   4.4 °C     0 °C   8.8 °C     n/d       -   
3   Bari  1988    Marzo       4   6.1 °C    -1 °C    12 °C       -    61 %   
4   Bari  1988    Marzo       5  10.2 °C     6 °C    12 °C    1 mm    81 %   
5   Bari  1988    Marzo       6  10.2 °C     9 °C    12 °C   35 mm    90 %   
6   Bari  1988    Marzo       7   8.3 °C     6 °C    10 °C    3 mm    91 %   
7   Bari  1988    Marzo       8   6.1 °C   4.6 °C     8 °C   14 mm    83 %   
8   Bari  1988    Marzo       9   3.5 °C     1 °C     7 °C   12 mm    82 %   
9   Bari  1988    Marzo      10   6.1 °C     3 °C     9 °C    2 mm    69 %   
10  Bari  1988    Marzo      11   5.1 °C   2.6 °C     9 °C     n/d    75 %   
11  Bari  1988    Marzo      12   8.7 °C     4 °C  14.6 °C     n/d       -   
12  Bari  1988    Marzo      13   9.5 °C     3 °C  16.2 °C       -    70 %   
13  Bari  1988    Marzo      14  11.1 °C     6 °C  18.2 °C       -    66 %   
14  Bari  1988    Marzo      15  10.3 °C     3 °C  15.2 °C       -    69 %   
15  Bari  1988    Marzo      16    15 °C    11 °C  21.2 °C       -    60 %   
16  Bari  1988    Marzo      17  12.5 °C     9 °C  16.6 °C     n/d    67 %   
17  Bari  1988    Marzo      18  10.9 °C     8 °C  13.8 °C    1 mm    73 %   
18  Bari  1988    Marzo      19   8.6 °C     6 °C  10.6 °C    5 mm    60 %   
19  Bari  1988    Marzo      20   7.6 °C   2.6 °C    11 °C       -    59 %   
20  Bari  1988    Marzo   

# Things to do: 
## analysis of the data - first of all to process them: 
*  ##  delete %, km/h etc. from the values in dataframes +  to transform strings into numerical values (floats) for appropriate variables, eg "6.6 °C" (string) into 6.6 (float)
* ## connected to the previous point, to replace missing values with monthly averages of the corresponding variable (or to drop the observation?)
* ## adding new data column to dataframes: daily range of temperature as Tmax-Tmin
* ## deleting unneccesary data columns from dataframes - Umidita, Raffica, Fenomeni,  Info (perhaps not really needed) 
* ## transliting Italian names of the variables (columns) into English, perhaps also name of the months inside of the dataframes
* ## then we can compute quarterly(?) (ie January to March, April to June...) means and variances for each city
